### ISSUE TITLE
[Doc]: kwargs in set_ticks when no labels are given

### ISSUE THREAD

**Issue by StefRe** [matplotlib/matplotlib#23272]
### Documentation Link

https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axis.Axis.set_ticks.html?highlight=set_ticks#matplotlib.axis.Axis.set_ticks

### Problem

If I want to set the tick label fontsize, I can do the following:
```
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.set_xticks((0,1,2,3), list('0123'), **{'fontsize': 20})
```
`ax.set_xticks((0,1,2,3), **{'fontsize': 20})` doesn't work as correctly documented:
> **kwargs
>     [Text](https://matplotlib.org/devdocs/api/text_api.html#matplotlib.text.Text) properties for the labels. These take effect only if you pass labels. In other cases, please use [tick_params](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.tick_params.html#matplotlib.axes.Axes.tick_params).

So far so good. But if I try `ax.tick_params('x', **{'fontsize': 20})` as suggested for this case, I get a `ValueError: keyword fontsize is not recognized`. 

A workaround is
```
for label in ax.get_xticklabels():
    label.set_fontsize(20)
```
or using pyplot: `plt.xticks(**{'fontsize': 20})`

### Suggested improvement

The description of `kwargs` should be changed to something like

> [Text](https://matplotlib.org/devdocs/api/text_api.html#matplotlib.text.Text) properties for the labels. These take effect only if you pass labels. In other cases, please use [tick_params](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.tick_params.html#matplotlib.axes.Axes.tick_params) for the parameters listed there. Other Text parameters can be changed by [setting](https://matplotlib.org/devdocs/api/text_api.html#matplotlib.text.Text.set) the Text properties of the ticklabels.

**Comment by StefRe** [comment]
> So you may as well write:

Ah, yes of course - sorry, I was just shortening down an example where I had a dict with multiple parameters.

> A problem is that `fontsize` is called `labelsize` in `tick_params`

OK, then let's take `fontweight` as an example which has not correspondence  in `tick_params`

**Comment by tacaswell** [comment]
I think the better fix here is to make sure that all of text properties can be passed through `tick_params` because we would like people to always use that rather than grabing the actual ticks to mutate them (which mostly works until it does not in very confusing action-at-a-distance ways due to caching).

However, short term, I agree with @StefRe that adjusting the docstring to be less expansive (but maybe not suggesting to grab the ticklabels!) is the easier fix.

**Comment by StefRe** [comment]
> ... rather than grabing the actual ticks to mutate them (which mostly works until it does not in very confusing action-at-a-distance ways due to caching).

@tacaswell  does this also pertain to using the pyplot interface (`plt.xticks(fontweight='bold')`) which uses `update`:
https://github.com/matplotlib/matplotlib/blob/2f5e7f31c384b266fa46e9abb0f7ac445d1848a0/lib/matplotlib/pyplot.py#L1819-L1821

in other words: would you rather refer to [`pyplot.xticks`](https://matplotlib.org/devdocs/api/_as_gen/matplotlib.pyplot.xticks.html) than to [`Text.set`](https://matplotlib.org/devdocs/api/text_api.html#matplotlib.text.Text.set)?

**Comment by StefRe** [comment]
>  we would like people to always use that rather than grabing the actual ticks to mutate them (which mostly works until it does not in very confusing action-at-a-distance ways due to caching).

although this approach is used in the official example [Centering labels between ticks](https://matplotlib.org/devdocs/gallery/ticks/centered_ticklabels.html)?

**Comment by oscargus** [comment]
Yes, `horizontalalignment` is another `Text` property that cannot be passed through `tick_params`, so currently the "bad" mutate approach is the only way to obtain it.

**Comment by timhoffm** [comment]
See https://github.com/matplotlib/matplotlib/issues/20644 for a discussion on alignment.

**Comment by anntzer** [comment]
I rediscovered this issue in #28456, requesting support for `ax.set_xticks(<no positions> **kwargs)`  I didn't notice that `plt.xticks` actually already implements this possibility.
On that thread @tacaswell argued (as above) to further expand the tick_params API to at least cover all the missing text properties.  I guess that'd work too, although I find that less discoverable than set_xticks: `tick_params` is a fairly idiosyncratic API which I personally have to re-google (well, really, grep the source ;-)) every time I use it whereas `set_xticks` is just "normal".

**Comment by timhoffm** [comment]
IIRC (I'm on mobile so cannot verify) there's a difference. 
kwargs only modify the existing tick instances. `tick_params()` in contrast also change the default for new ticks (ticks are mostly reused, but not guaranteed to be persistent, e.g. you may need additional ticks later). See the warning in the kwargs description of https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.set_xticklabels.html#matplotlib.axes.Axes.set_xticklabels. I believe this also applies for `plt.xticks()` and `set_ticks()`.

**Comment by tacaswell** [comment]
I'm a bit worried about having two ways to do the same thing.

This does point to a gap in our APIs in that we have ways to control the tick locations and the values are formatted to strings, but we do not have an analogous machinery for setting the styling of the ticks conditional on the location (e.g. "I want to make odd valued labels red and even value labels bold" or "I want to alternate the label color between black and grey starting with the first tick being black").

**Comment by anntzer** [comment]
> I'm a bit worried about having two ways to do the same thing.

It's probably not going to happen, but I guess if we only wanted one way to do things, `set_xticks(**kwargs)` actually looks more like the "normal" API and I would soft-deprecate `tick_params()`...

> This does point to a gap in our APIs in that we have ways to control the tick locations and the values are formatted to strings, but we do not have an analogous machinery for setting the styling of the ticks conditional on the location (e.g. "I want to make odd valued labels red and even value labels bold" or "I want to alternate the label color between black and grey starting with the first tick being black").

I guess this would likely belong to Formatter API, e.g. allowing Formatter.format_ticks to return not a list of strings but a list of dicts, which are Text properties for the corresponding ticks?

**Comment by timhoffm** [comment]
There's a conflict of interest: In a world, where ticks can change e.g. because of zooming, you can't have both options to be available:
- I want all ticks to be formatted consistently
- I want the second ticks (or the tick at position 5.5) to be red and bold.

Well, in theory you could, but that would require a ticking system that is much more complex than ours.

---

More broadly speaking, 99% of the users just need the "consistent formatting" case. That should be easy and the default. A design where each tick carries separate full style information is not well suited for this. I still believe considering all ticks on an axis as one collective entity would be the right way to go (performance wise, but also in terms of the right level of abstraction), and [we might be able to get there in a backward-compatible way](https://github.com/matplotlib/matplotlib/issues/5665#issuecomment-875155264).

I believe such a system would be sufficient for 99.99% of users: Consistent formatting for static and dynamic cases, *and* + free formatting for static cases.
When such a level of abstraction is in place, you could theoretically extend it with index and/or position dependent formatting rules (e.g. "format all negative ticks red" or Tom's suggestions). But TBH, that'd be more to show off rather than a really needed feature. IMHO the first step moving ticks to a collective entity is the important one. It should improve performance and make internal tick handling much simpler.

**Comment by anntzer** [comment]
> There's a conflict of interest: In a world, where ticks can change e.g. because of zooming, you can't have both options to be available:
> 
> * I want all ticks to be formatted consistently
> * I want the second ticks (or the tick at position 5.5) to be red and bold.
> 
> Well, in theory you could, but that would require a ticking system that is much more complex than ours.

I *think* this would be a matter of adding support for Formatter subclasses like
```python
clas MyCustomFormatter(Formatter):
    def format_ticks(self, values):  # Perhaps a new API name, although that seems unnecessary
        # Perhaps replace str(val) by something around the right super() call.
        return [{"text": str(val), "color": "red" if i == 1 else "black"} for i, val in enumerate(values)]
        # or
        return [{"text": str(val), "color": "red" if val == 5.5 else "black"} for val in values]
```
(and possibly providing an easier way to construct such formatters like we currently do for FuncFormatter), which doesn't seem too horrible.

**Comment by timhoffm** [comment]
> I _think_ this would be a matter of adding support for Formatter subclasses

Maybeee. This would be a paradigm shift. Currently, formatters do only format the text, i.e. which characters are used. They do not participate in styling. Styling is currently done via a combination of `tick_params` as global style settings and the attributes of the individual ticks. Adding formatters to the mix will make consistent handling more complex.

I think, also style formatters would require to first go into the direction of https://github.com/matplotlib/matplotlib/issues/23272#issuecomment-2189832658. When the state of individual ticks is gone, it will be much simpler to compose the desired style for the ticks on demand.

**Comment by tacaswell** [comment]
We did propose the tick refactor as part of the work in the next NASA grant : 
https://github.com/matplotlib/2024-NASA-ROSES-F7/blob/cf5c987b3bcaf455d0902a8d431e6946c63e312b/proposal/mpl_cartopy.tex#L955-L993  (this went in about a month and a half ago, no idea when we will hear back).
